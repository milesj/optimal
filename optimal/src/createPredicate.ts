import { AnySchema, InferSchemaType, Predicate } from './types';

/**
 * Create a reusable predicate from a schema that validates a value to
 * return a boolean if no validation errors arise. If null or undefined
 * is returned, will return `false`, regardless of nullable or undefinable
 * state.
 */
export function createPredicate<S extends AnySchema, T = InferSchemaType<S>>(
	schema: S,
): Predicate<T> {
	return (value) => {
		try {
			if (value == null) {
				throw new Error('Avoid null/undefined');
			}

			schema.validate(value);
		} catch {
			return false;
		}

		return true;
	};
}
