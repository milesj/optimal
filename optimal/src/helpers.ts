import {
	Constructor,
	DefaultValue,
	DefaultValueInitializer,
	Schema,
	SchemaValidateOptions,
	UnknownObject,
} from './types';
import { ValidationError } from './ValidationError';

export function isObject(value: unknown): value is object {
	return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
	return (
		isObject(value) &&
		typeof (value as UnknownObject).schema === 'function' &&
		typeof (value as UnknownObject).state === 'function' &&
		typeof (value as UnknownObject).type === 'function' &&
		typeof (value as UnknownObject).validate === 'function'
	);
}

export function isValidDate(value: Date): value is Date {
	return (
		Object.prototype.toString.call(value) === '[object Date]' &&
		!Number.isNaN(value.getTime()) &&
		value.toString() !== 'Invalid Date'
	);
}

export function isValidNumber(value: unknown): value is number {
	return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
	return typeof value === 'string' && value !== '';
}

export function createArray(value: unknown): unknown[] {
	if (value === undefined) {
		return [];
	}

	return Array.isArray(value) ? [...(value as unknown[])] : [value];
}

export function createDate(value: unknown): Date {
	if (value instanceof Date) {
		return value;
	}

	if (value === undefined || value === null) {
		return new Date();
	}

	return new Date(value as number);
}

export function createObject<T = UnknownObject>(value: unknown) {
	return (isObject(value) ? { ...value } : {}) as unknown as T;
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(object: unknown, contract: Constructor<T>): object is T {
	if (!object || typeof object !== 'object') {
		return false;
	}

	if (object instanceof contract) {
		return true;
	}

	let current = object;

	while (current) {
		if (current.constructor.name === 'Object') {
			return false;
		}

		if (
			current.constructor.name === contract.name ||
			(current instanceof Error && current.name === contract.name)
		) {
			return true;
		}

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		current = Object.getPrototypeOf(current);
	}

	return false;
}

export function invalid(
	condition: boolean,
	message: string,
	path: string = '',
	value: unknown = undefined,
) {
	if (condition) {
		return;
	}

	throw new ValidationError(message, path, value);
}

export function invariant(condition: boolean, message: string) {
	if (condition) {
		return;
	}

	throw new Error(message);
}

export function logUnknown(unknownFields: object, pathPrefix?: string) {
	const unknownKeys = Object.keys(unknownFields);

	if (unknownKeys.length > 0) {
		throw new Error(
			`${pathPrefix ? `Unknown "${pathPrefix}" fields` : 'Unknown fields'}: ${unknownKeys.join(
				', ',
			)}.`,
		);
	}
}

export function pathKey(path: string): string {
	if (path.endsWith(']')) {
		const index = path.lastIndexOf('[');

		return index > 0 ? path.slice(index) : path;
	}

	const index = path.lastIndexOf('.');

	return index > 0 ? path.slice(index + 1) : path;
}

export function prettyValue(value: unknown): string | null {
	switch (typeof value) {
		case 'string':
			return `"${value}"`;

		case 'number':
		case 'function':
			return String(value);

		case 'object': {
			if (value === null) {
				return `\`null\``;
			}

			if (value.constructor !== Object) {
				return value.constructor.name === 'Array' ? null : `\`${value.constructor.name}\``;
			}

			return null;
		}

		default:
			return `\`${value}\``;
	}
}

export function tryAndCollect(
	validator: () => boolean | void,
	validError: ValidationError,
	collectErrors?: boolean,
): boolean {
	let result = false;

	try {
		const value = validator();

		if (typeof value === 'boolean') {
			result = value;
		}
	} catch (error: unknown) {
		if (error instanceof Error && collectErrors) {
			validError.addError(error);
		} else {
			throw error;
		}
	}

	return result;
}

export function extractDefaultValue<T>(
	defaultValue: DefaultValue<T>,
	path: string,
	{ currentObject, rootObject }: SchemaValidateOptions,
) {
	return typeof defaultValue === 'function'
		? (defaultValue as DefaultValueInitializer<T>)(path, currentObject!, rootObject!)
		: defaultValue;
}
