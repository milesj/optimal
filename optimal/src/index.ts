/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import * as schemas from './schemas';

export type Schemas = typeof schemas;

export { schemas };
export * from './createPredicate';
export * from './createSchema';
export * from './criteria';
export * from './optimal';
export * from './OptimalError';
export * from './schemas/array';
export * from './schemas/blueprint';
export * from './schemas/bool';
export * from './schemas/custom';
export * from './schemas/date';
export * from './schemas/func';
export * from './schemas/instance';
export * from './schemas/lazy';
export * from './schemas/number';
export * from './schemas/object';
export * from './schemas/regex';
export * from './schemas/schema';
export * from './schemas/shape';
export * from './schemas/string';
export * from './schemas/tuple';
export * from './schemas/union';
export * from './types';
export * from './ValidationError';
