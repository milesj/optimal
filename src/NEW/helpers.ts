import { Constructor, UnknownObject, Schema } from './types';

export function createDate(value: unknown): Date {
  return value instanceof Date ? value : new Date(value as number);
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(
  object: unknown,
  contract: Constructor<T> | Function,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      return false;
    }

    if (
      current.constructor.name === contract.name ||
      (current instanceof Error && current.name === contract.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}

export function invariant(condition: boolean, message: string, path: string = '') {
  if (__DEV__) {
    if (condition) {
      return;
    }

    // TODO no errorPrefix
    const prefix = path ? `Invalid field "${path}".` : '';

    throw new Error(`${prefix} ${message}`.trim());
  }
}

export function isObject(value: unknown): value is object {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
  return (
    isObject(value) &&
    typeof (value as UnknownObject).validate === 'function' &&
    typeof (value as UnknownObject).type === 'function'
  );
}

export function isValidDate(value: Date): value is Date {
  return (
    Object.prototype.toString.call(value) === '[object Date]' &&
    !Number.isNaN(value.getTime()) &&
    value.toString() !== 'Invalid Date'
  );
}

export function isValidNumber(value: unknown): value is number {
  return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export function pathKey(path: string): string {
  const index = path.lastIndexOf('.');

  return index > 0 ? path.slice(index + 1) : path;
}
