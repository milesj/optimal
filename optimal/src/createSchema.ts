import { invalid, tryAndCollect } from './helpers';
import { OptimalError } from './OptimalError';
import {
	AnySchema,
	Criteria,
	CriteriaFactory,
	DefaultValueInitializer,
	InferSchemaType,
	Schema,
	SchemaOptions,
	SchemaState,
	SchemaValidateOptions,
} from './types';

/**
 * Run all validation checks that have been enqueued and return a type casted value.
 * If a value is undefined, inherit the default value, else throw if required.
 * If nullable and the value is null, return early.
 */
function validate<T>(
	state: SchemaState<T>,
	validators: Criteria<T>[],
	initialValue: unknown,
	path: string = '',
	{
		collectErrors = true,
		currentObject = {},
		rootObject = currentObject,
	}: SchemaValidateOptions = {},
): T | null {
	const { defaultValue, metadata } = state;

	let value: unknown = initialValue;

	// Handle undefined
	if (value === undefined) {
		value =
			typeof defaultValue === 'function'
				? (defaultValue as DefaultValueInitializer<T>)(path, currentObject, rootObject)
				: defaultValue;

		invalid(!state.required, 'Field is required and must be defined.', path);
	} else {
		if (__DEV__ && metadata.deprecatedMessage) {
			// eslint-disable-next-line no-console
			console.info(`Field "${path}" is deprecated. ${metadata.deprecatedMessage}`);
		}

		invalid(!state.never, 'Field should never be used.', path);
	}

	// Handle null
	if (value === null) {
		invalid(state.nullable, 'Null is not allowed.', path);
	}

	// Run validations and produce a new value
	const optimalError = new OptimalError();

	validators.forEach((test) => {
		if (
			(test.skipIfNull && value === null) ||
			(test.skipIfOptional && !state.required && value === state.defaultValue)
		) {
			return;
		}

		tryAndCollect(
			() => {
				const result = test.validate(value as T, path, {
					collectErrors,
					currentObject,
					rootObject,
				});

				if (result !== undefined) {
					value = result as T;
				}
			},
			optimalError,
			collectErrors,
		);
	});

	if (optimalError.errors.length > 0) {
		throw optimalError;
	}

	return value as T;
}

export function createSchema<S extends AnySchema, T = InferSchemaType<S>>(
	{ api, cast, defaultValue, type }: SchemaOptions<T>,
	criteria: (Criteria<T> | CriteriaFactory<T>)[] = [],
): S {
	const state: SchemaState<T> = {
		defaultValue,
		metadata: {},
		never: false,
		nullable: false,
		required: false,
		type,
	};

	const validators: Criteria<T>[] = [];

	criteria.forEach((crit) => {
		if (typeof crit === 'function') {
			const validator = crit(state);

			if (validator) {
				validators.push(validator);
			}
		} else {
			validators.push(crit);
		}
	});

	const schema: Schema<T> = {
		schema() {
			return type;
		},
		type() {
			return state.type;
		},
		validate(value, path, options) {
			const result = validate(state, validators, value, path, options)!;

			return cast && result !== null ? cast(result) : result;
		},
	};

	Object.entries(api).forEach(([name, method]) => {
		Object.defineProperty(schema, name, {
			enumerable: true,
			// Create a new schema so that our chainable API is immutable
			value: (...args: unknown[]) =>
				createSchema({ api, cast, defaultValue, type }, [
					...criteria,
					(nextState) => method(nextState, ...args),
				]),
		});
	});

	return schema as S;
}
