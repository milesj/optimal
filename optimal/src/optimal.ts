import { isObject } from './helpers';
import { OptimalError } from './OptimalError';
import { shape } from './schemas/shape';
import { Blueprint, DeepPartial, SchemaValidateOptions, UnknownObject } from './types';
import { ValidationError } from './ValidationError';

export interface OptimalOptions {
	file?: string;
	name?: string;
	prefix?: string;
	unknown?: boolean;
}

export interface Optimal<T extends object> {
	configure: (options: OptimalOptions) => void;
	validate: (struct: DeepPartial<T>, options?: SchemaValidateOptions) => Required<T>;
}

export function optimal<Schemas extends object>(
	blueprint: Blueprint<Schemas>,
	baseOpts: OptimalOptions = {},
): Optimal<Schemas> {
	const options: OptimalOptions = {};
	let schema = shape(blueprint);

	function configure(nextOpts: OptimalOptions) {
		if (!isObject(nextOpts)) {
			throw new TypeError('Optimal options must be a plain object.');
		}

		Object.assign(options, nextOpts);
		schema = schema.exact(!options.unknown);
	}

	configure(baseOpts);

	return {
		configure,
		validate(struct, validateOptions) {
			const object = struct as UnknownObject;

			try {
				return schema.validate(struct, options.prefix ?? '', {
					...validateOptions,
					currentObject: object,
					rootObject: object,
				});
			} catch (error: unknown) {
				let invalid: OptimalError;
				let prefix = '';

				if (error instanceof OptimalError) {
					invalid = error;
				} else {
					invalid = new OptimalError();
					invalid.addError(error as ValidationError);
				}

				if (options.name) {
					invalid.schema = options.name;
					prefix = options.name;
				}

				if (options.file) {
					invalid.file = options.file;
					prefix = prefix ? `${prefix} (${options.file})` : options.file;
				}

				if (prefix) {
					invalid.message = `${prefix}: ${invalid.message}`;
				}

				throw invalid;
			}
		},
	};
}
