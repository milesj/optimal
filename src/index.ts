/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal from './optimal';
import Builder, { custom, func } from './Builder';
import ArrayBuilder, { array } from './ArrayBuilder';
import BooleanBuilder, { bool } from './BooleanBuilder';
import InstanceBuilder, { builder, instance, date, regex } from './InstanceBuilder';
import NumberBuilder, { number } from './NumberBuilder';
import ObjectBuilder, { object, blueprint } from './ObjectBuilder';
import ShapeBuilder, { shape } from './ShapeBuilder';
import StringBuilder, { string } from './StringBuilder';
import TupleBuilder, { tuple } from './TupleBuilder';
import UnionBuilder, { union } from './UnionBuilder';

export const predicates = {
  array,
  blueprint,
  bool,
  builder,
  custom,
  date,
  func,
  instance,
  number,
  object,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  array,
  blueprint,
  bool,
  builder,
  custom,
  date,
  func,
  instance,
  number,
  object,
  regex,
  shape,
  string,
  tuple,
  union,
};

export {
  Builder,
  ArrayBuilder,
  BooleanBuilder,
  InstanceBuilder,
  NumberBuilder,
  ObjectBuilder,
  ShapeBuilder,
  StringBuilder,
  TupleBuilder,
  UnionBuilder,
};

export type Predicates = typeof predicates;

export * from './types';

export default optimal;
