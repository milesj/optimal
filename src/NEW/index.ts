/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import * as criteria from './criteria';
import createSchema from './createSchema';
import { array } from './schemas/array';
import { bool } from './schemas/bool';
import { instance, date, regex } from './schemas/instance';
import { number } from './schemas/number';
import { object } from './schemas/object';
import { schema, blueprint } from './schemas/schema';
import { string } from './schemas/string';

export { criteria, createSchema };

export const predicates = {
  array,
  blueprint,
  bool,
  // custom,
  date,
  // func,
  instance,
  number,
  object,
  regex,
  schema,
  // shape,
  string,
  // tuple,
  // union,
};

export * from './schemas/array';
export * from './schemas/bool';
export * from './schemas/instance';
export * from './schemas/number';
export * from './schemas/schema';
export * from './schemas/string';
export * from './schemas/object';
export * from './types';
