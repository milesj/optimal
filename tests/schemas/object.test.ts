import { object, blueprint, string, number, ObjectSchema } from '../../src/NEW';
import { runChecks, runInProd } from '../helpers';

describe('object()', () => {
  let schema: ObjectSchema<Record<string, string>>;

  beforeEach(() => {
    schema = object().of(string());
  });

  it('errors if a non-predicate is passed', () => {
    expect(() => {
      // @ts-expect-error Allow non-predicate
      schema = object(123);
    }).toThrowErrorMatchingSnapshot();
  });

  it('doesnt error if a predicate is not passed', () => {
    expect(() => {
      object();
    }).not.toThrow();
  });

  it('doesnt error if a predicate is passed', () => {
    expect(() => {
      schema = object().of(string());
    }).not.toThrow();
  });

  it('returns an empty object for no data', () => {
    expect(runChecks(schema)).toEqual({});
  });

  it('errors if a non-object is passed', () => {
    expect(() => {
      runChecks(
        schema,
        // @ts-expect-error Test invalid type
        'foo',
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if a non-object is passed, when not using a predicate', () => {
    expect(() => {
      runChecks(
        object(),
        // @ts-expect-error Test invalid type
        'foo',
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('returns the type alias when no contents', () => {
    expect(object().type()).toBe('object');
  });

  it('returns the type alias with contents type', () => {
    expect(object().of(number()).type()).toBe('object<number>');
  });

  it('returns default value if value is undefined', () => {
    expect(runChecks(object({ foo: 'foo' }).of(string()))).toEqual({ foo: 'foo' });
  });

  // it('returns default value from factory if value is undefined', () => {
  //   expect(
  //     runChecks(
  //       object(string(), () => ({ foo: 'bar' })),
  //       undefined,
  //       {
  //         struct: { multiplier: 3 },
  //       },
  //     ),
  //   ).toEqual({ foo: 'bar' });
  // });

  it('checks each item in the object', () => {
    expect(() => {
      runChecks(schema, {
        a: 'foo',
        b: 'bar',
        // @ts-expect-error Test invalid type
        c: 123,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if an object item is invalid; persists path with index', () => {
    expect(() => {
      runChecks(schema, {
        // @ts-expect-error Test invalid type
        foo: 123,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it('supports objects of objects', () => {
    const nestedPredicate = object().of(object().of(string()));

    const data = {
      a: {
        foo: '123',
        bar: '456',
      },
      b: {
        baz: '789',
      },
    };

    expect(runChecks(nestedPredicate, data)).toEqual(data);
  });

  it('errors correctly for objects in objects', () => {
    const nestedPredicate = object().of(object().of(string()));

    expect(() => {
      runChecks(nestedPredicate, {
        a: {
          foo: '123',
          // @ts-expect-error Test invalid type
          bar: 456,
        },
        b: {
          baz: '789',
        },
      });
    }).toThrowErrorMatchingSnapshot();
  });

  describe('production', () => {
    it(
      'returns an empty object if value is object',
      runInProd(() => {
        expect(runChecks(schema, {})).toEqual({});
      }),
    );

    it(
      'returns default value if value is undefined',
      runInProd(() => {
        const def = { foo: 'foo' };

        schema = object(def);

        expect(runChecks(schema)).toEqual(def);
      }),
    );

    // it(
    //   'returns default value from factory if value is undefined',
    //   runInProd(() => {
    //     expect(runChecks(object(string(), () => ({ foo: 'foo' })))).toEqual({ foo: 'foo' });
    //   }),
    // );

    it(
      'bypasses checks and returns value',
      runInProd(() => {
        expect(
          runChecks(
            schema,
            // @ts-expect-error Test invalid type
            { foo: 123 },
          ),
        ).toEqual({ foo: '123' });
      }),
    );
  });

  describe('notEmpty()', () => {
    beforeEach(() => {
      schema.notEmpty();
    });

    it('errors if value is empty', () => {
      expect(() => {
        runChecks(schema, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if not empty', () => {
      expect(() => {
        runChecks(schema, { foo: '123' });
      }).not.toThrow();
    });

    it('doesnt error if null', () => {
      expect(() => {
        schema.nullable();

        runChecks(schema, null);
      }).not.toThrow();
    });
  });

  describe('sizeOf()', () => {
    beforeEach(() => {
      schema.sizeOf(3);
    });

    it('errors if length doesnt match', () => {
      expect(() => {
        runChecks(schema, {});
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error if length matches', () => {
      expect(() => {
        runChecks(schema, { a: '1', b: '2', c: '3' });
      }).not.toThrow();
    });

    it('doesnt error if null', () => {
      expect(() => {
        schema.nullable();

        runChecks(schema, null);
      }).not.toThrow();
    });
  });
});

describe('blueprint()', () => {
  it('errors if a non-predicate is passed', () => {
    expect(() => {
      runChecks(
        blueprint(),
        // @ts-expect-error Allow invalid type
        { value: 123 },
      );
    }).toThrow('Invalid field "key.value". Must be an instance of "Predicate".');
  });
});
