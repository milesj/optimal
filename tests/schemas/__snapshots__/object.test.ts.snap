// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object() checks each item in the object 1`] = `"Invalid field \\"key.c\\". Must be a string."`;

exports[`object() errors correctly for objects in objects 1`] = `"Invalid field \\"key.a.bar\\". Must be a string."`;

exports[`object() errors if a non-object is passed 1`] = `"Invalid field \\"key\\". Must be a plain object."`;

exports[`object() errors if a non-object is passed, when not using a predicate 1`] = `"Invalid field \\"key\\". Must be a plain object."`;

exports[`object() errors if a non-predicate is passed 1`] = `"A schema blueprint is required for object values."`;

exports[`object() errors if an object item is invalid; persists path with index 1`] = `"Invalid field \\"key.foo\\". Must be a string."`;

exports[`object() notEmpty() errors if value is empty 1`] = `"Invalid field \\"key\\". Object cannot be empty."`;

exports[`object() sizeOf() errors if length doesnt match 1`] = `"Invalid field \\"key\\". Object must have 3 properties."`;
